File: client/src/app/containers/token/Token.js
// @flow weak

import { connect }            from 'react-redux';
import { bindActionCreators } from 'redux';
import gql                    from 'graphql-tag';
import { compose, graphql }   from 'react-apollo';
import { Token }              from '../../views'
import * as globalsActions    from '../../redux/modules/globals';
import * as plotsActions      from '../../redux/modules/plots';
import * as tokenActions      from '../../redux/modules/token';

/* -----------------------------------------
  GraphQL - Apollo client
 ------------------------------------------*/

const query = gql`
  query (
    $identifier: String!,
    $mentionSubredditPlotRange: String,
    $mentionTotalPlotRange: String,
  ) {
    token: tokenByIdentifier(identifier: $identifier) {
      id
      shortName
      longName
      imageUrl
      website
      priceUSD
      priceBTC
      volumeUSD24h
      volumeBTC24h
      marketCapUSD
      marketCapBTC
      availableSupply
      totalSupply
      maxSupply
      percentChange1h
      percentChange24h
      percentChange7d

      mentionTotalCounts(timeRange: $mentionTotalPlotRange) {
        count
        timestamp
      }

      subreddits {
        id
        displayName
        name
        imageUrl
      }

      subredditMentions(timeRange: $mentionSubredditPlotRange) {
        subreddit {
          id
          displayName
        }

        mentionTotalCounts {
          count
          timestamp
        }
      }
    }
  }
`;
const queryOptions = {
  options: ({
    match,
    mentionSubredditPlotRange,
    mentionTotalPlotRange,
  }) => {
    return {
      variables: {
        identifier: match.params.identifier,
        mentionSubredditPlotRange: mentionSubredditPlotRange,
        mentionTotalPlotRange: mentionTotalPlotRange,
      },
    };
  },
};

/* -----------------------------------------
  Redux
 ------------------------------------------*/

const mapStateToProps = (state) => {
  return {
    currency: state.globals.currency,
    mentionSubredditPlotRange: state.plots.mentionSubredditPlotRange,
    mentionTotalPlotRange: state.plots.mentionTotalPlotRange,
    nightMode: state.globals.nightMode,
    pricePlotRange: state.plots.pricePlotRange,
    scrollActive: state.globals.scrollActive,
    selectedTicker: state.token.selectedTicker,
  };
};

const mapDispatchToProps = (dispatch) => {
  return bindActionCreators(
  {
    changeMentionSubredditPlotRange: plotsActions.changeMentionSubredditPlotRange,
    changeMentionTotalPlotRange: plotsActions.changeMentionTotalPlotRange,
    changePricePlotRange: plotsActions.changePricePlotRange,
    changeScrollActive: globalsActions.changeScrollActive,
    changeSelectedTicker: tokenActions.changeSelectedTicker,
  },
    dispatch
  );
};

export default compose(
  graphql(query, queryOptions),
  connect(mapStateToProps, mapDispatchToProps)
)(Token);

File: server/app/models/token.rb
# == Schema Information
#
# Table name: tokens
#
#  id                 :integer          not null, primary key
#  short_name         :string           not null
#  long_name          :string           not null
#  created_at         :datetime         not null
#  updated_at         :datetime         not null
#  image_url          :string           default(""), not null
#  website            :string           default(""), not null
#  price_usd          :decimal(, )      default(0.0), not null
#  price_btc          :decimal(, )      default(0.0), not null
#  volume_usd_24h     :decimal(, )      default(0.0), not null
#  market_cap_usd     :decimal(, )      default(0.0), not null
#  available_supply   :decimal(, )      default(0.0), not null
#  total_supply       :decimal(, )      default(0.0), not null
#  max_supply         :decimal(, )      default(0.0), not null
#  percent_change_1h  :decimal(, )      default(0.0), not null
#  percent_change_24h :decimal(, )      default(0.0), not null
#  percent_change_7d  :decimal(, )      default(0.0), not null
#
# Indexes
#
#  index_tokens_on_short_name  (short_name) UNIQUE
#

class Token < ApplicationRecord
  has_many :keywords
  has_many :markets
  has_many :mention_counts, through: :keywords
  has_many :subreddit_tokens
  has_many :subreddits, through: :subreddit_tokens

  validates :short_name, uniqueness: true
  validates :long_name, uniqueness: true
end
